cli/command/trust/key_generate.go

```diff
@@ -0,0 +1,116 @@
+package trust
+
+import (
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/trust"
+	"github.com/docker/notary"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/data"
+	tufutils "github.com/docker/notary/tuf/utils"
+	"github.com/spf13/cobra"
+)
+
+func newKeyGenerateCommand(dockerCli command.Streams) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "key-generate NAME [NAME...]",
+		Short: "Generate a signing key",
+		Args:  cli.RequiresMinArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return generateKeys(dockerCli, args)
+		},
+	}
+	return cmd
+}
+
+// key names can use alphanumeric + _ + - characters
+var validKeyName = regexp.MustCompile(`^[a-zA-Z0-9\_\-]+$`).MatchString
+
+// validate that all of the key names are unique and are alphanumeric + _ + -
+// and that we do not already have public key files in the current dir on disk
+func validateKeyArgs(keyNames []string, cwdPath string) error {
+	uniqueKeyNames := map[string]struct{}{}
+	for _, keyName := range keyNames {
+		if !validKeyName(keyName) {
+			return fmt.Errorf("key name \"%s\" must not contain special characters", keyName)
+		}
+
+		if _, ok := uniqueKeyNames[keyName]; ok {
+			return fmt.Errorf("key names must be unique, found duplicate key name: \"%s\"", keyName)
+		}
+		uniqueKeyNames[keyName] = struct{}{}
+
+		pubKeyFileName := keyName + ".pub"
+		if _, err := os.Stat(filepath.Join(cwdPath, pubKeyFileName)); err == nil {
+			return fmt.Errorf("public key file already exists: \"%s\"", pubKeyFileName)
+		}
+	}
+	return nil
+}
+
+func generateKeys(streams command.Streams, keyNames []string) error {
+	var genKeyErrs []string
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	if err := validateKeyArgs(keyNames, cwd); err != nil {
+		return err
+	}
+	for _, keyName := range keyNames {
+		fmt.Fprintf(streams.Out(), "\nGenerating key for %s...\n", keyName)
+		freshPassRet := trust.GetPassphraseRetriever(streams.In(), streams.Out())
+		if err := generateKey(keyName, cwd, trust.GetTrustDirectory(), freshPassRet); err != nil {
+			fmt.Fprintf(streams.Out(), err.Error())
+			genKeyErrs = append(genKeyErrs, keyName)
+		} else {
+			pubFileName := strings.Join([]string{keyName, "pub"}, ".")
+			fmt.Fprintf(streams.Out(), "Successfully generated and loaded private key. Corresponding public key available: %s\n", pubFileName)
```

Comment:
As discussed offline - it might be worth noting when we write the docs that if one uses `key-generate`, they don't need to `key-load` after